packages/core/src/extensions/Blocks/PreviousBlockTypePlugin.ts

```diff
@@ -47,152 +43,133 @@ export const PreviousBlockTypePlugin = () => {
       init() {
         return {
           prevBlockAttrs: {} as any,
-          needsUpdate: false,
+          updatedBlocks: new Set<string>(),
+          prevTransactionOldAttrs: {} as any,
         };
       },
 
       apply(transaction, prev, oldState, newState) {
-        prev.needsUpdate = false;
-        prev.prevBlockAttrs = {};
+        prev.updatedBlocks.clear();
+
         if (!transaction.docChanged || oldState.doc.eq(newState.doc)) {
           return prev;
         }
 
-        const transform = combineTransactionSteps(oldState.doc, [transaction]);
-        // const { mapping } = transform;
-        const changes = getChangedRanges(transform);
-
-        // TODO: instead of iterating through the entire document, only check nodes affected by the transactions
```

Comment:
I agree it would make for a good optimization, though the reason I removed the TODO is because I don't think it would quite work with the way list item animations function. Take the following example editor content:

1. List Item 1
2. List Item 2
3. List Item 3

After changing the type of List Item 1 to a paragraph, the content becomes:

List Item 1
1. List Item 2
2. List Item 3

So even though List Item 2 and List Item 3 may not have been affected by any number of previous transactions, they still get updated.

I think one way to go around this issue is to store the previous transaction ranges in the state, kind of how we do with block attributes before the previous transaction (`prevTransactionOldAttrs`). Then we would go over the previous transaction's ranges as well as the current transaction's, if it's generated by the list indexing plugin.

Reply:
Good notes! Let's revisit when we want to optimise this